src/main/scala/io/iohk/ethereum/blockchain/sync/FastSync.scala

```diff
@@ -4,804 +4,473 @@ import java.time.Instant
-          syncState = syncState.copy(pendingMptNodes = Seq(StateMptNodeHash(syncState.targetBlock.stateRoot)))
-        } else {
-          syncState = syncState.updateTargetBlock(targetBlockHeader, syncConfig.fastSyncBlockValidationX, updateFailures = false)
-          log.info(s"Changing target block to ${targetBlockHeader.number}, new safe target is ${syncState.safeDownloadTarget}")
-        }
-
-      case LastBlockValidationFailed =>
-        log.info(s"Changing target block after failure, to ${targetBlockHeader.number}, new safe target is ${syncState.safeDownloadTarget}")
-        syncState = syncState.updateTargetBlock(targetBlockHeader, syncConfig.fastSyncBlockValidationX, updateFailures = true)
-    }
-
-    private def removeRequestHandler(handler: ActorRef): Unit = {
-      context unwatch handler
-      assignedHandlers -= handler
-    }
-
-    private def discardLastBlocks(startBlock: BigInt, blocksToDiscard: Int): Unit = {
-      (startBlock to ((startBlock - blocksToDiscard) max 1) by -1).foreach { n =>
-        blockchain.getBlockHeaderByNumber(n).foreach { headerToRemove =>
-          blockchain.removeBlock(headerToRemove.hash, withState = false)
-        }
-      }
-      appStateStorage.putBestBlockNumber((startBlock - blocksToDiscard - 1) max 0)
-    }
+    private def updateBestBlockIfNeeded(receivedHashes: Seq[ByteString]): Unit = {
+      val headers = for {
+        hash   <- receivedHashes
+        header <- blockchain.getBlockHeaderByHash(hash)
+        _      <- blockchain.getBlockBodyByHash(hash)
+        _      <- blockchain.getReceiptsByHash(hash)
+      } yield header
 
-    @tailrec
-    private def processHeaders(peer: Peer, headers: Seq[BlockHeader]): HeaderProcessingResult = {
       if (headers.nonEmpty) {
-        val header = headers.head
-        processHeader(header, peer) match {
-          case Left(result)        => result
-          case Right(headerAndDif) =>
-            updateSyncState(headerAndDif._1, headerAndDif._2)
-            if (header.number == syncState.safeDownloadTarget){
-              ImportedTargetBlock
-            } else {
-              processHeaders(peer, headers.tail)
-            }
-        }
-      } else
-        HeadersProcessingFinished
-    }
-
-    private def validateHeader(header: BlockHeader, peer: Peer): Either[HeaderProcessingResult, BlockHeader] = {
-      val shouldValidate = header.number >= syncState.nextBlockToFullyValidate
-
-      if (shouldValidate) {
-        validators.blockHeaderValidator.validate(header, blockchain.getBlockHeaderByHash) match {
-          case Right(_) =>
-            updateValidationState(header)
-            Right(header)
-
-          case Left(error) =>
-            log.warning(s"Block header validation failed during fast sync at block ${header.number}: $error")
-            Left(ValidationFailed(header, peer))
-        }
-      } else {
-        Right(header)
+        val bestReceivedBlock = headers.maxBy(_.number).number
```

Comment:
Wouldn't be `headers.map(_.number).max` a bit easier to read?

Reply:
but would it be faster?